File: api/app/schemas/transaction.py
import re
from datetime import UTC, datetime
from pydantic import BaseModel, Field, field_validator, model_validator

UUID_PATTERN = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$",
)

class TransactionCreateRequest(BaseModel):
    kind: str = Field(pattern="^(income|expense|transfer|credit_charge|credit_payment)$")
    amount: float = Field(gt=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    description: str | None = Field(default=None, max_length=255)
    occurred_at: datetime | None = Field(
        default_factory=lambda: datetime.now(UTC),
    )
    category_id: str | None = None
    cash_wallet_id: str | None = None
    bank_account_id: str | None = None
    credit_card_id: str | None = None
    target_cash_wallet_id: str | None = None
    target_bank_account_id: str | None = None

    @field_validator("currency")
    @classmethod
    def normalize_currency(cls, value: str) -> str:
        return value.upper()

    @field_validator(
        "category_id",
        "cash_wallet_id",
        "bank_account_id",
        "credit_card_id",
        "target_cash_wallet_id",
        "target_bank_account_id",
    )
    @classmethod
    def validate_uuid_like(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not UUID_PATTERN.fullmatch(value):
            raise ValueError("must be a valid UUID string.")
        return value

    @model_validator(mode="after")
    def validate_shape(self) -> "TransactionCreateRequest":
        asset_count = int(self.cash_wallet_id is not None) + int(
            self.bank_account_id is not None,
        )
        target_asset_count = int(self.target_cash_wallet_id is not None) + int(
            self.target_bank_account_id is not None,
        )
        has_credit_card = self.credit_card_id is not None

        if self.kind == "income":
            if asset_count != 1 or has_credit_card or target_asset_count != 0:
                raise ValueError(
                    "income requires exactly one source asset and no credit card/target assets.",
                )
            return self

        if self.kind == "expense":
            valid_expense_shape = (
                (asset_count == 1 and not has_credit_card)
                or (asset_count == 0 and has_credit_card)
            )
            if not valid_expense_shape or target_asset_count != 0:
                raise ValueError(
                    "expense must use exactly one source asset OR one credit card, without target assets.",
                )
            return self

        if self.kind == "transfer":
            if asset_count != 1 or target_asset_count != 1 or has_credit_card:
                raise ValueError(
                    "transfer requires exactly one source asset and one target asset.",
                )
            return self

        if self.kind == "credit_charge":
            if asset_count != 0 or target_asset_count != 0 or not has_credit_card:
                raise ValueError(
                    "credit_charge requires credit_card_id and no asset references.",
                )
            return self

        if self.kind == "credit_payment":
            if asset_count != 1 or target_asset_count != 0 or not has_credit_card:
                raise ValueError(
                    "credit_payment requires one source asset and one credit_card_id.",
                )
            return self

        return self

class TransactionUpdateRequest(BaseModel):
    kind: str | None = Field(default=None, pattern="^(income|expense|transfer|credit_charge|credit_payment)$")
    amount: float | None = Field(default=None, gt=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    description: str | None = Field(default=None, max_length=255)
    occurred_at: datetime | None = None
    category_id: str | None = None
    cash_wallet_id: str | None = None
    bank_account_id: str | None = None
    credit_card_id: str | None = None
    target_cash_wallet_id: str | None = None
    target_bank_account_id: str | None = None

    @field_validator("currency")
    @classmethod
    def normalize_currency(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return value.upper()

    @field_validator(
        "category_id",
        "cash_wallet_id",
        "bank_account_id",
        "credit_card_id",
        "target_cash_wallet_id",
        "target_bank_account_id",
    )
    @classmethod
    def validate_uuid_like(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not UUID_PATTERN.fullmatch(value):
            raise ValueError("must be a valid UUID string.")
        return value

class TransactionResponse(BaseModel):
    id: str
    kind: str
    amount: float
    currency: str
    description: str | None = None
    occurred_at: datetime
    category_id: str | None = None
    cash_wallet_id: str | None = None
    bank_account_id: str | None = None
    credit_card_id: str | None = None
    target_cash_wallet_id: str | None = None
    target_bank_account_id: str | None = None
    created_at: datetime
    updated_at: datetime

File: mobile/lib/features/transactions/models/transaction.dart
enum TransactionKind { income, expense, transfer, creditCharge, creditPayment }

class Transaction {
  const Transaction({
    required this.id,
    required this.kind,
    required this.amount,
    required this.currency,
    this.description,
    required this.occurredAt,
    this.categoryId,
    this.cashWalletId,
    this.bankAccountId,
    this.creditCardId,
    this.targetCashWalletId,
    this.targetBankAccountId,
  });

  final String id;
  final TransactionKind kind;
  final double amount;
  final String currency;
  final String? description;
  final DateTime occurredAt;
  final String? categoryId;
  final String? cashWalletId;
  final String? bankAccountId;
  final String? creditCardId;
  final String? targetCashWalletId;
  final String? targetBankAccountId;

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      kind: _parseKind(json['kind'] as String),
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] as String,
      description: json['description'] as String?,
      occurredAt: DateTime.parse(json['occurred_at'] as String),
      categoryId: json['category_id'] as String?,
      cashWalletId: json['cash_wallet_id'] as String?,
      bankAccountId: json['bank_account_id'] as String?,
      creditCardId: json['credit_card_id'] as String?,
      targetCashWalletId: json['target_cash_wallet_id'] as String?,
      targetBankAccountId: json['target_bank_account_id'] as String?,
    );
  }

  static TransactionKind _parseKind(String kind) {
    switch (kind.toLowerCase()) {
      case 'income':
        return TransactionKind.income;
      case 'expense':
        return TransactionKind.expense;
      case 'transfer':
        return TransactionKind.transfer;
      case 'credit_charge':
        return TransactionKind.creditCharge;
      case 'credit_payment':
        return TransactionKind.creditPayment;
      default:
        return TransactionKind.expense;
    }
  }
}
